src/inputs.rs

```diff
@@ -71,8 +71,17 @@ impl PartInput {
         worker_count: WorkerCount,
         mut resume_state: StepStateBytes,
     ) -> PyResult<(PartBundle, TotalPartCount)> {
-        let mut keys: Vec<StateKey> = self.0.call_method0(py, "list_parts")?.extract(py)?;
-        keys.sort();
+        let keys = {
+            let mut keys = self
+                .0
+                .call_method0(py, "list_parts")?
+                .as_ref(py)
+                .iter()?
+                .map(|i| i.and_then(PyAny::extract::<StateKey>))
+                .collect::<PyResult<Vec<StateKey>>>()?;
+            keys.sort();
+            keys
+        };
```

Comment:
you could collect to a `BTreeSet` to avoid the sort, the rest of the code should work as it is (if we don't expect duplicate keys):
```rust
        let keys = self
            .0
            .call_method0(py, "list_parts")?
            .as_ref(py)
            .iter()?
            .map(|i| i.and_then(PyAny::extract::<StateKey>))
            .collect::<PyResult<BTreeSet<StateKey>>>()?;
```

Reply:
Also, not related to this PR, but I noticed this while reviewing, at the end of this function we do:
```rust
let remaining_keys: Vec<StateKey> = resume_state.into_keys().into_iter().collect();
assert!(remaining_keys.is_empty(), "Partition keys in resume state that are not in partition list; changing partition counts? recovery state routing bug?");
```

Given that we drop `resume_state` anyway at the end of the function, we could proxy the `is_empty` method of the hashmap in `StepStateBytes`:
```rust
impl StepStateBytes {
    pub(crate) fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}
```
And then do:
```rust
assert!(resume_state.is_empty(), "Partition keys in resume state that are not in partition list; changing partition counts? recovery state routing bug?");
```

Reply:
Great suggestions, thank you.